shuts down all contexts on sigint; remove g_sigint_gc

koonpeng opened the pull request:
This changes the SIGINT handler to only run when no other handlers are installed, it also shuts down all contexts instead of just the default context.

This also fix a bug where a context is not properly shutdown in the sigint handler, with this bug fixed there should no longer be a need to explicitly call `process.exit` in the handler, we will let node exit gracefully when all work in the event loop are finished.

Discussion:
I make the signal handler run only when no other handlers are installed, I did this because maybe an application would want to handle SIGINT differently and not shutdown rclnodejs. But doing this has the side effect that if the app installs a signal handler to clean up it's own resource and it expects rclnodejs to cleanup itself (because that is the behavior when no handlers is installed), rclnodejs's handler would not run.

So given this scenario, I wonder if it is better to
* Run the handler only when it is the only one installed, has the side effect mentioned above.
* Run the handler regardless, apps that wants it's own SIGINT behavior that does not clean things up would have to remove rclnodejs's handler.
* Do not install signal handlers. Users may be confused why ctrl+c does not work by default.

felixdivo commented:
See also #721, it is very much related

minggangw commented on index.js:
```diff
@@ -407,12 +407,6 @@ let rcl = {
 };
 
-process.on('SIGINT', () => {
-  debug('Catch ctrl+c event and will cleanup and terminate.');
-  rcl.shutdown();
-  process.exit(0);
```
The initial purpose of installing a handler and handling SIGINT here is that I once found the program cannot be terminated by `ctrl+c` sometimes and this gives a chance to quit the process. I'm not sure if we may still meet the issue.

minggangw replied:
Finally, I remember the reason why we need to shut down here (considering the code was written several years ago...).

At that time, there is no kind of Context concept, so `rcl.shutdown()` will stop the sub-thread and the main thread could quit as expected. But now user may create more than one Context, which means more sub-threads are running simultaneously, so simply executing `rcl.shutdown()` is not enough and we must ensure all contexts are cleaned up properly when terminating.
https://github.com/RobotWebTools/rclnodejs/blob/d41f7e91510cd55ff672dad8f7f0be095ed8c00c/index.js#L305

Commit:
add package-lock to .gitignore

minggangw commented:
Discussion about SIGINT https://github.com/RobotWebTools/rclnodejs/pull/721#issuecomment-726650101

Commit:
update ecma version to recognize spread operator

Commit:
add tests for signal handler

Commit:
remove g_sigint_gc; sigint handler shuts down all contexts; only run sigint handler when no longer handlers are installed

felixdivo commented:
I think we should not ignore `package-lock.json`;). As you said, controversial. But it's AFAIK best practice, and [the manual](https://github.com/npm/cli/blob/release-6.14.7/docs/content/configuring-npm/package-lock-json.md) says:

> This file is intended to be committed into source repositories, and serves various purposes:
>
> - Describe a single representation of a dependency tree such that teammates, deployments, and continuous integration are guaranteed to install exactly the same dependencies.
>
>- Provide a facility for users to "time-travel" to previous states of node_modules without having to commit the directory itself.
>
>- To facilitate greater visibility of tree changes through readable source control diffs.
>
>- And optimize the installation process by allowing npm to skip repeated metadata resolutions for previously-installed packages.

felixdivo commented on index.js:
```diff
@@ -407,11 +407,16 @@ let rcl = {
-  rcl.shutdown();
-  process.exit(0);
-});
+  // only run when no other signal handlers are installed
```
I really like this change and it should replace #721 for these lines. Two things though: The debug call in line 411 should be moved down into the if-check, and maybe we should document how this works in the module documentation? E.g something along:

> Only if no other handlers for `SIGINT` events are registered, *rclnodejs* will shut down all live contexts in these events. This is useful to gracefully exit the process when `CTRL+C` is pressed. Once other signal handlers are registered using `process.on('SIGINT', some_handler)`, the handler of *rclnodejs* will not attempt to shut down the contexts.

felixdivo replied:
Alternatively, we could supply an extra argument to specify `auto_cleanup=False` to `rclnodejs.init(conetxt, argv)` to make it explicit. This might cause less confusion for users.

Commit:
make the default behavior to shut down all contexts; add a method to remove the default signal handler

Commit:
Merge remote-tracking branch 'origin/develop' into no-signal-handler

Commit:
only shutdown if context is running, avoiding the most common exception thrown

koonpeng commented:
Thanks for the comments. After some thoughts, I changed to default behavior to always cleanup rclnodejs. Reason is because I figured most apps would fall into one of these

* Does not have any long running background tasks, expects ctrl+c to "just work" to gracefully exit the program.
* Has some background tasks, so it installs some signal handlers to clean up it's resources.
* Has some special behavior when it receive SIGINT, it may only want to perform cleanup when some conditions are met.

The first 2 is probably the most common and the new behavior should work for them. Particularly for case 2 because we are no longer forcing node to exit, it should work well with the app's signal handler. For case 3, I added a new function `rclnodejs.removeSignalHandlers` to allow the app to perform it's own cleanup logic.

koonpeng commented:
> I think we should not ignore package-lock.json;). As you said, controversial. But it's AFAIK best practice, and the manual says:

This is a complicated topic, the docs recommend committing package-lock.json but there are some arguments which suggests why we should ignore it. See https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi. Since package-lock is not committed in this repo, I assumed that is what rclnodejs prefer. Perhaps @minggangw can help explain if we should/should not ignored it?

Commit:
fix tests

minggangw commented:
I haven't met any problem when running `npm install` without package-lock.json until now.The [npm doc](https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json) said this file should be committed into the repo, but I think if everything runs well without it, we could ignore it now and add it back if something goes wrong in the future.

felixdivo commented:
> See https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi.

Interesting point. Thank's for sharing. I agree with you now. Delete it and wait what happens. Maybe add the link to that article to the `.gitignore` file or something like that.

The npm docs should better discuss that...

felixdivo commented on index.js:
```diff
@@ -409,13 +409,31 @@ let rcl = {
+   * @returns {undefined}
+   */
+  removeSignalHandlers() {
+    process.removeListener('SIGINT', _sigHandler);
```
```suggestion
    // this will not throw even if the handler is already removed
    process.removeListener('SIGINT', _sigHandler);
```

I tested it with a generic `EventHandler`.

felixdivo commented on index.js:
```diff
@@ -409,13 +409,31 @@ let rcl = {
-  process.exit(0);
-});
+const _sigHandler = () => {
+  // shuts down all live contexts. Applications which wishes to use their own signal handlers
```
```suggestion
  // shuts down all live contexts. Applications that wish to use their own signal handlers
```

felixdivo commented on index.js:
```diff
@@ -409,13 +409,31 @@ let rcl = {
+  // should call `rclnodejs.removeSignalHandlers`.
+  debug('Catch ctrl+c event and will cleanup and terminate.');
+  for (let ctx of rcl._contextToNodeArrayMap.keys()) {
+    if (!rcl.isShutdown(ctx)) {
+      rcl.shutdown(ctx);
+    }
```
```suggestion
    rcl.shutdown(ctx);
```

This check is already done within `rcl.shutdown()` since #721.

Commit:
remove unneeded check

Commit:
improve comments

Commit:
skip sigint test for windows

Commit:
trigger ci

koonpeng commented:
Windows CI is failing sporadically, I wonder if it's because it's using node v12.19.0 which has the segfault bug. Also the cpp build failed, not sure if it's related.

minggangw reviewed (approved):
LGTM! Thanks for making this improvement and I think the handling of sigint is perfect now :)

minggangw commented:
```bash
tep 17/21 : RUN rosdep install --from-paths $ROS2_WS/ros2-linux/share --ignore-src --rosdistro foxy -y --skip-keys "console_bridge fastcdr fastrtps osrf_testing_tools_cpp poco_vendor rmw_connext_cpp rosidl_typesupport_connext_c rosidl_typesupport_connext_cpp rti-connext-dds-5.3.1 tinyxml_vendor tinyxml2_vendor urdfdom urdfdom_headers"
 ---> Running in 00f69051d49d
ERROR: the following packages/stacks could not have their rosdep keys resolved
to system dependencies:
rqt_bag: Cannot locate rosdep definition for [rosbag2_transport_py]
```
There is an error on travis-ci now, I tried to fix it by #727, but it seems not workable. So, I'd like merge this PR in advance and get back to recheck travis-ci with the next package of https://ci.ros2.org/view/packaging/job/packaging_linux/lastSuccessfulBuild/, thanks!
